### DISCUSSION TITLE
FTP server feature for accessing embedded and SD card file system

### DISCUSSION THREAD

**Discussion by nikvostro** [arendst/Tasmota#20074] [Ideas]
Hi team!

Love your project! Thanks a lot for your efforts in development and troubleshooting! You're doing a great job! :)

Nowadays there is a demand has appeared for me to have a convinient access to internal filesystem of ESP32. It's more actual for accessing attached SD card. So accessing FS via Web UI isn't really comfortable you know... For managing 1-5-10 files OK, but for bigger amount it becomes a challenge. I can't say this will be a routine work for me but sometimes I need to manage moving some files in/out ESP32.

Have you discussed FTP server implementation for managing FS before? Is it feasible with little efforts?

p.s. I've tried seaching for keywords "ftp server" in discussions/issues etc. on Git and Discord channecls - nothing there.

Nikolay

**Comment by sfromis** [comment]
If someone wants to code a FTP server on ESP32, it would be quite feasible with the Berry language and some work. However, I get by without even wanting this, and without doing this via the rudimentary web UI.

One approach is using http and tools like curl. Should also work on ESP8266 devices.
`http://mytasmotadevice/ufsd?download=/myfile.txt`
`curl --form "ufsu=@/home/mypath/myfile.txt" "http://mytasmotadevice/ufsu"`

The Tasmota command `UrlFetch` can also work great for some scenarios, depending on how the process is to be controlled.

**Reply by nikvostro** [reply]
@sfromis hi! 

Thank you for sharing these commands!

This can be automated with some script on PC indeed. But in my case this looks inconvinient because I'm managing to browse folders first to check their contents. Sure I can parse `http://mytasmotadevice/ufsd?download=/` page to view a list... but this approach looks a bit strange.

My case: I'm wokring on simple non-AI voice assistant based on [https://www.elechouse.com/product/simplevr-speaker-independent-voice-recognition-module/](url). For repsonses/confirmations from ESP32 I'm using a set of MP3s pre-uploaded to SD card which I'm updating from time to time. Sometimes in a bulk.... This makes me crazy :)

p.s. in my home automation everything works decentralized (except Zigbee gateway) on Scripting and Rules only. I find migrating  to Berry too complicated for my solution. So in my view FTP server should be running independently of scripting. As a driver (like HTTP API) + commands to manage it.

p.s.p.s. by the way, could you tell me if "[fo](https://tasmota.github.io/docs/Scripting-Language/#sensor-logging:~:text=in%20root%20directory-,fr%3Dfo(%22/%22%200),-for%20cnt%201)" command works with SD card for listing folder contents? I've failed to realize that :(  For UFS it worked for me but not for SDC. Going to raise an Issue soon

**Reply by sfromis** [reply]
Using scripting and rules is where it get complex to manage. You have much better options by getting started with Berry.

Here's a simple way of getting a listing of file names using Berry, via a http request:
`http://mytasmotadevice/cm?cmnd=Br+import+path+return+path.listdir('/').concat(',')`
The usage of `+` is not Berry, but URL escapes (For readability, I'm using `+` instead of `%20`).
The command also works fine via MQTT. The path for a SD card would be `/sd/` instead of just `/`.

When it comes to the scripting feature, I'm not using that, and have no answers about getting things to work there. I've found Berry preferable, and easier to deal with.

And in open source projects, features tend to be added by submissions, not by "I want to have **this**".

**Reply by nikvostro** [reply]
@sfromis sure, I do understand the policy of open projects. My initial inquiry was just about asking if this was discsussed befrore already and if it feasible to implement but not about impelemting for me personally. As an **Idea** in general. I've just described my use case. Thank you for sharing another one micro-script and your experince. Will be thinking of Berry...

**Reply by s-hadinger** [reply]
I'm not a big fan of FTP, because the protocol is outdated and not very convinient to implement. I understand however that for transfering a large number of files, it can be handy. I tried implementing WebDAV a long time ago, to leverage file transfer over http, but must client were very buggy, and it ended to be unsable.

Let me check if there is an easy way to add FTP with an existing lib. This one looks promising.
https://github.com/xreef/SimpleFTPServer

**Reply by nikvostro** [reply]
@s-hadinger hi! Actually any of protocols with folder/files listing fetaure for easier (direct let's say) file management would work: FTP, WebDAV, SSH etc... If it's possible to integrate with minimal efforts it would be great. But again - not sure if this feature will become really desirable for others :)

**Comment by Jason2866** [comment]
A FTP server will not integrated in Tasmota. If needed "someone" has to do in Berry.
EDIT: Seems your lucky day ;-)

**Reply by s-hadinger** [reply]
Actually not so much. I tried, listing of directory works, but downloading or uploading is stuck. I have no clue why and will unfortunately not spend time to debug it.
Sorry

**Reply by nikvostro** [reply]
Thank you for your efforts!
Maybe today isn't the best day for debug and there is a chance to try later? Or somebody would participate further... 

@Jason2866 1:0

**Reply by gemu2015** [reply]
@nikvostro 
i have implemented FTP now. if works on ESP32 and ESP8266. you may give it a try with my fork.
#define USE_FTP
#define USER_FTP              "user"
#define PW_FTP                "password"

must use passive FTP mode

**Reply by sfromis** [reply]
To be clear, that implementation is only if using the scripter feature, not Tasmota in general.

**Reply by gemu2015** [reply]
no, has nothing to doo with scripting. its part of file system (xdrv_50_filesystem.ino)

**Reply by Jason2866** [reply]
@nikvostro Have you tested the [PR](https://github.com/arendst/Tasmota/pull/20402) providing a FTP Server?

**Reply by gemu2015** [reply]
@Jason2866 
i made extensive tests with esp8266 and esp32. it works very stable, but a little slow, about 70 kb per second

**Reply by Jason2866** [reply]
PR is merged

**Reply by nikvostro** [reply]
Great news! Thanks a lot!
To be honest I'm not monitoring a PRs. Will do my best to try soon
